funcx_sdk/funcx/sdk/executor.py

```diff
@@ -315,12 +315,22 @@ def event_loop_thread(self, eventloop):
         eventloop.run_until_complete(self.web_socket_poller())
 
     async def web_socket_poller(self):
-        # TODO: if WebSocket connection fails, we should either retry connecting and
-        # backoff or we should set an exception to all of the outstanding futures
-        await self.ws_handler.init_ws(start_message_handlers=False)
-        await self.ws_handler.handle_incoming(
-            self._function_future_map, auto_close=True
-        )
+        """Start ws and listen for tasks.
+        If a remote disconnect breaks the ws, close the ws and reconnect"""
+        while True:
+            await self.ws_handler.init_ws(start_message_handlers=False)
+            status = await self.ws_handler.handle_incoming(
+                self._function_future_map, auto_close=True
+            )
+            if status is False:
+                # handle_incoming broke from a remote side disconnect
+                # we should close and re-connect
+                log.info("Attempting ws close")
+                await self.ws_handler.close()
+                log.info("Attempting ws re-connect")
+            else:
+                # clean exit
+                break
```

Comment:
Given that the `WebSocketPollingTask` is used in the client directly as well, I believe that it, not the executor, should be responsible for the retries. I haven't had time to read the usages very closely yet, so please correct me if I'm wrong.

Reply:
@sirosen the client instantiates a `WebSocketPollingTask` when it runs with `asynchronous=True`. I placed the logic for reconnects here because this is one level above any component that touches websockets, so it should be safe to terminate the connections from here.

Reply:
This means that the websockets usage within a `FuncXClient` is not getting automatic retries.

I'm looking at how the `WebSocketPollingTask` in the client gets initialized, and it has `auto_start=True`. That ultimately kicks off an instance of `handle_incoming` and `send_outgoing`.
For those async tasks, if the connection closes there will be no reconnect.

Given that `WebSocketPollingTask` is meant to be able to run async tasks in the background, I'm not sure why we would _want_ reconnection logic to be here one level above it? Is there an advantage to this, rather than having the `WebSocketPollingTask` itself handle reconnection?

You mention being able to terminate connections. Perhaps, rather than closing and reopening, the better model is that the poller connects and reconnects as necessary, but that its main loops are `while not self.closed` rather than `while True`?

Reply:
@sirosen The `FuncXClient` in `asynchronous` interacts with the `WebSocketPollingTask` quite differently, so I wasn't sure if having the `handle_incoming` call retry makes sense. I'll update here after taking a closer look.